pkg/controllers/routereflector/ctrl_ops.go

```diff
@@ -0,0 +1,295 @@
+func (c *ctrl) revertFailedModifications() error {
+	// Iterating on all known nodes
+	for _, kubeNode := range c.kubeNodes {
+		status, ok := c.routeReflectorsUnderOperation[kubeNode.GetUID()]
+		if !ok {
+			continue
+		}
+
+		rrRevertCounter.WithLabelValues(kubeNode.GetName()).Inc()
+
+		// Node was under operation, better to revert it
+		var err error
+		if status {
+			// Previous operation was add
+			err = c.datastore.RemoveRRStatus(kubeNode, c.findCalicoNode(kubeNode))
+		} else {
+			// Previous operaion was remove
+			err = c.datastore.AddRRStatus(kubeNode, c.findCalicoNode(kubeNode))
+		}
+		if err != nil {
+			log.Errorf("Failed to revert node %s: %s", kubeNode.GetName(), err.Error())
+			return err
+		}
+
+		// Update Kubernetes node
+		if _, err := c.k8sNodeClient.Update(context.Background(), kubeNode, metav1.UpdateOptions{}); err != nil {
+			log.Errorf("Failed to revert update node %s: %s", kubeNode.GetName(), err.Error())
+			return err
+		}
+
+		delete(c.routeReflectorsUnderOperation, kubeNode.GetUID())
+	}
+
+	return nil
+}
+
+// updateNodeLabels calculates topology and updates all the necessary labels in datastore
+func (c *ctrl) updateNodeLabels(affectedNodes map[*corev1.Node]bool) error {
+	// on case of a zone has less nodes then required, count missing
+	missingRouteReflectors := 0
+
+	// Calculate topology
+	for _, status := range c.topology.GetRouteReflectorStatuses(affectedNodes) {
+		// Increment the expected with the number of missing
+		status.ExpectedRRs += missingRouteReflectors
+		log.Infof("Route refletor status in zone(s) %v (actual/expected) = %d/%d", status.Zones, status.ActualRRs, status.ExpectedRRs)
+
+		for _, n := range status.Nodes {
+			n := c.kubeNodes[n.GetUID()]
+
+			if !affectedNodes[n] {
+				continue
+			} else if status.ExpectedRRs == status.ActualRRs {
+				break
+			}
+
+			// Calculate diff to detect scale operation
+			diff := status.ExpectedRRs - status.ActualRRs
+			if diff == 0 {
+				continue
+			}
+
+			// Update labels on node
+			if updated, err := c.updateRRStatus(n, diff); err != nil {
+				log.Errorf("Unable to update node %s: %s", n.GetName(), err.Error())
+				return err
+			} else if updated && diff > 0 {
+				status.ActualRRs++
+			} else if updated && diff < 0 {
+				status.ActualRRs--
+			}
+
+		}
+
+		// Recalculate missing nodes
+		if status.ExpectedRRs != status.ActualRRs {
+			log.Infof("Actual number %d is different than expected %d", status.ActualRRs, status.ExpectedRRs)
+			missingRouteReflectors = status.ExpectedRRs - status.ActualRRs
+		}
+	}
+
+	// There are still missing route reflectors
+	if missingRouteReflectors != 0 {
+		rrMissingGauge.Set(float64(missingRouteReflectors))
+		log.Errorf("Actual number is different than expected, missing: %d", missingRouteReflectors)
+	}
+
+	return nil
+}
+
+// updateBGPTopology creates/updates BGP peer configurations based on topology
+func (c *ctrl) updateBGPTopology(affectedNodes map[*corev1.Node]bool) error {
```

Comment:
By default I don't recommend to edit BGP peers manually, because the automation will delete any modification. And you are right the code doesn't deal with this scenario. The reason is; to deal with this edge case needs to increase the complexity of code, because we don't have the Kubernetes node object at this point which triggers the entire reconcile.
I haven't found easy solution for this so left it as it is, because user musn't edit BGP peers if operator has enabled.